### DISCUSSION TITLE
What's the best way to invalidate a nested list?

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2517] [Q&A] [answered]
I have a quick question. What is the best practice to invalidate `venue.bankAccounts`?

I have a query like this:

```graphql
query VenueBankAccount($venueId: String!) {
  venue(id: $venueId) {
    id 
    bankAccounts { id }
  }
}
```

When I call `createBankAccount(venueId, bankAccount)` successfully, I want to refetch that venue's list of bank accounts. I don't want an optimistic mutation since this should be correct data.

What is the best way to do this? Can I call something like `cache.invalidate()` specifically on the bank accounts of that venue? I don't need to invalidate the entire `Venue` – only `Venue.${id}.bankAccounts`.

The docs point to this example: https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#invalidating-individual-fields

However, this seems to invalidate fields on `todos`, rather than something like `user.${id}.todos`.

Thanks!

**Comment by JoviDeCroock** [comment, marked as answer]
You should be able to do `cache.invalidate({ __typename: 'Venue', id: args.venueId })` or `cache.invalidate({ __typename: 'Venue', id: args.venueId }, 'bankAccounts')` and it should start refetching, that will invalidate the venue with said id.

**Reply by nandorojo** [reply]
Got it, the second one seems to be what I want. 

And to clarify, if I call this inside of `updates.Mutation.createBankAccount()`, it will get called _after_ a successful creation of the bank account?

```ts
setDefaultBankAccount(parent, args, cache, info) {
  const __typename: Venue['__typename'] = 'Venue'
  const field: keyof Venue = 'bankAccounts'
  
  cache.invalidate({
    __typename,
    id: args.venueId
  }, field)
}
```

Does that look right?

I probably don't need to revalidate the entire venue, since only the bank accounts will update.

**Reply by JoviDeCroock** [reply]
> And to clarify, if I call this inside of updates.Mutation.createBankAccount(), it will get called after a successful creation of the bank account?

Yes, and the code looks all right to me

**Reply by nandorojo** [reply]
Okay, great! Thanks again.

So if I want to invalidate both `bankAccounts` and `defaultBankAccount`, I would just call `cache.invalidate` twice:

```ts
setDefaultBankAccount(parent, args, cache, info) {
  const __typename: Venue['__typename'] = 'Venue'
  
  cache.invalidate({
    __typename,
    id: args.venueId
  }, 'bankAccounts')

  cache.invalidate({
    __typename,
    id: args.venueId
  }, 'defaultBankAccount')
}
```

**Reply by JoviDeCroock** [reply]
Yes or invalidate the entire venue 😅 both will work

**Reply by nandorojo** [reply]
> invalidate the entire venue 😅

Fair...lol. Will this "delete" the venue from the cache, or just refetch it? 

PS I'm migrating my entire app from SWR + REST to URQL + GQL so thanks for all the answers here – life-saving.

**Reply by JoviDeCroock** [reply]
`invalidate` does a delete, so the result becomes null. You can also update to just append it to the list of accounts but yes, invalidating is often a good shout imo

**Reply by nandorojo** [reply]
Hmm, interesting. I don't exactly want that, since it would make the UI flicker by hiding the venue temporarily. I'd prefer to refetch while preserving the cache in the meantime.

SWR has a `revalidate()` function for this usecase. Is it possible to do something like `cache.revalidate({ __typename, id })`, or do I need to do a full `cache.query()`? 

I find `{ __typename, id }` to be a nicer API for this use case, over creating a full query string for every relevant query.

**Reply by JoviDeCroock** [reply]
Currently we don't offer `revalidate` but what you can do is

```
const bankAccounts = cache.resolve({ __typename, id: args.venueId }, 'bankAccounts');
cache.link({ __typename, id: args.venueId }, 'bankAccounts', [...bankAccounts, cache.keyOfEntity({ __typename: 'BankAccount', id: args.bankAccount })])
```

Which will link it to the existing bankAccount or the one contained in the resulting selection of your mutation, if you don't have access to that our current advise is invalidating. You can look at https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/#partial-results if the UI being in a loading state is a problem

**Reply by nandorojo** [reply]
I don't mind the UI being a loading state. I just don't want the stale data to get deleted and for the user to see an empty Venue after having fetched it.

However, I just realized that, while `cache.invalidate` deletes the object from the cache, it won't get deleted from the hooks that have already fetched it. Is that correct? 

I'm now seeing the benefits of URQL's hook-level caching behavior in action 😎.

So in summary, `cache.invalidate` should be fine. I'll try this all out. Thanks so much Jovi!

**Comment by kitten** [comment]
Yes,

make sure to pass the keyavle object (i.e. either at least `{ __typename, id}` or as per the `keys` config if you've got that for the parent type for the object the field is on, and the field's name as the second argument as a string. The third argument is for field arguments, which you don't seem to have any in your example.

https://formidable.com/open-source/urql/docs/api/graphcache/#invalidate

The first argument may also be the output of `cache.resolve` as most other cache methods if that's easier, since it can return keys.

And alternatively, you can also delete the venue by resolving it, which would avoid you having to delete only the field and would instead delete the venue object.

It's all described on that docs link

**Reply by nandorojo** [reply]
Thanks once again for the useful info.

Is there an equivalent to `cache.invalidate` that refetches without deleting? Something like `cache.revalidate({ __typename, id })`?

Update: I think I answered my own question [here](https://github.com/FormidableLabs/urql/discussions/2517#discussioncomment-3005552).

My assumption is that, while `cache.invalidate({ __typename: 'Venue', id })` removes the `Venue.id` from the cache, the `useVenueQuery` hook itself will still preserve the stale state, so there won't be a UI flicker of empty data.

Update 2:

Looks like `cache.invalidate` is exactly what I needed.
